### PR TITLE
AP_RangeFinder: implement VL53L0X rangefinder

### PR DESCRIPTION
**Pull request by tridge**
very short range, useful for small vehicles

### CODE REVIEW

**Comment by mirkix** [comment]
@tridge Tested with [BBBmini](https://github.com/mirkix/BBBMINI) . Unfortunately BeagleBone based boards use often I2C bus 2, so it would be nice to check on I2C bus 2 too.

**Comment by tridge** [comment]
what we need is a hal.i2c.get_bus_list()
btw, on PX4 we always map internal i2c bus to 0 and external to 1. On some STM32 boards we use hw bus 2 as well, but it is mapped in the HAL

**Comment by patrickpoirier51** [comment]
Nice, really like this little device, guess you are aware of this: ArduPilot/ardupilot#5605
The next generation will offer faster speed and greater range http://www.st.com/content/st_com/en/products/imaging-and-photonics-solutions/proximity-sensors/vl53l1.html

Question: It seem that you are using continuous scan mode only? If that is the case, the range is limited to 1,2Meter. On single shot mode, we can get 1,8 Meter

**Comment by rmackay9** [comment]
Looks like many of function names are camel case?  I think we let that go in the EKF but not normally outside the EKF.

**Comment by patrickpoirier51** [comment]
@OXINARF 
Concerning I2C , allow me to diverge and ask you to add this one in your list : 
https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_RangeFinder/RangeFinder.cpp#L615

**Comment by OXINARF** [comment]
@patrickpoirier51 This issue affect all external I2C devices, so, once we fix it, we can fix all libraries.

**Comment by Kiwa21** [comment]
First of all, thanks @tridge for implementing this. I was planning to use the VL53L0X for the very last part of vertical landing. It's a tiny sensor that can be mounted even on landing gear !  I hope this will give me a strong reliable data to be able to hover or landing (correct climb_rate data) smoothly at very low altitude because even the data from SF11/C can't be trusted under 1 m.

Secondly, I very much agree " hal.i2c.get_bus_list()". Especially with PH2, I2C_2 is on the internal bus 0 while a lot of library are only searching on bus 1. For example, I had to do a poorly coded fix for this in the AP_IR_Lock_I2C.cpp to enable Pixy on I2C_2

`dev0 = std::move(hal.i2c_mgr->get_device(0, IRLOCK_I2C_ADDRESS));
	
	if(dev0){
		dev = std::move(hal.i2c_mgr->get_device(0, IRLOCK_I2C_ADDRESS));
	}else{
		dev = std::move(hal.i2c_mgr->get_device(1, IRLOCK_I2C_ADDRESS));
	}`

**Comment by mirkix** [comment]
@tridge Rebased it to master and tested with pixracer successfully. Would be nice if it will be merged.

**Comment by patrickpoirier51** [comment]
@tridge  tested ok but limited to 1,2M.
Could you consider switching to one shot -long range = 1,8M

Same library but call a different set of parameters, look here:
https://github.com/pololu/vl53l0x-arduino/blob/master/examples/Single/Single.ino
have to set #define LONG_RANGE on line 23

File: libraries/AP_RangeFinder/RangeFinder.cpp [hunk 1]
```diff
@@ -284,7 +285,11 @@ const AP_Param::GroupInfo RangeFinder::var_info[] = {
     // @Param: 3_TYPE
     // @DisplayName: Third Rangefinder type
     // @Description: What type of rangefinder device that is connected
+<<<<<<< 1c8d169814979b667f2ee3480d7f2d49198f3871
```
Comment (khancyr) [inline libraries/AP_RangeFinder/RangeFinder.cpp]: woups failed rebase ^^

**Comment by OXINARF** [comment]
Merged, thanks!

@patrickpoirier51 I think we don't want to do that by default as according to the comments there it reduces reliability.

**Comment by patrickpoirier51** [comment]
Fair enough,  better be safe than sorry ;-) , Thanks Francisco

**Pull request merged**